File: dci-adk-sample/src/main/java/com/domainmodeling/dci/adk/sample/java/moneytransfer/MoneyTransfer.java
package com.domainmodeling.dci.adk.sample.java.moneytransfer;

import adk.context.DCIContext;
import adk.exception.DCIRoleInstanceBuildException;
import lombok.AllArgsConstructor;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;

@AllArgsConstructor
public class MoneyTransfer implements DCIContext {

    private final String from;
    private final String to;

    public List<AccountLedger> transfer(BigDecimal amount) throws DCIRoleInstanceBuildException {

        Account fromAccount = ImmutableAccount.builder()
                .balance(BigDecimal.valueOf(100))
                .name(from)
                .build();

        Account toAccount = ImmutableAccount.builder()
                .balance(BigDecimal.valueOf(100))
                .name(to)
                .build();

        List<AccountLedger> ledgers = Arrays.asList(
                fromAccount.play(AccountTransferer.class).outcome(amount),
                toAccount.play(AccountTransferer.class).income(amount)
        );

        return ledgers;
    }
}

File: dci-adk-sample/src/main/kotlin/com/domainmodeling/dci/adk/sample/kotlin/moneytransfer/MoneyTransfer.kt
package com.domainmodeling.dci.adk.sample.kotlin.moneytransfer

import java.math.BigDecimal

class MoneyTransfer(private val fromAccount: String, private val toAccount: String) :
        AccountTransferer {
    fun transfer(transferAmount: BigDecimal): Array<AccountLedger> {

        val from = Account(
                BigDecimal(100.00),
                fromAccount
        );
        val to = Account(
                BigDecimal(100.00),
                toAccount
        );

        val ledgers: Array<AccountLedger> = arrayOf(
            from.outcome(transferAmount),
            to.income(transferAmount)
        );

        return ledgers;
    }
}